Fix adventure border rendering

a1exsh opened the pull request:
Some screen resolutions where the width or height are not a multiple of
`TILEWIDTH` (32 pixels) require special treatment: we pad the difference with
a repeatable pattern.

The pattern to take depends on Evil/Good interface selection: the same
coordinates don't always work for both.

Evil also has special bottom border where the usual approach doesn't give
perfect results.  This is also addressed now.

Additionally, this makes the bottom right (and middle) border junction
rendering correct.

Fix #4192

a1exsh commented:
Screenshots:

720x480:
![Screenshot from 2021-09-12 18-57-32](https://user-images.githubusercontent.com/489601/132996175-0ad799ef-894c-4e38-8b2c-e25572a2131a.png)
![Screenshot from 2021-09-12 18-58-44](https://user-images.githubusercontent.com/489601/132996203-c5a5bd85-df73-4d15-8a84-4049231a9f66.png)

800x600:
![Screenshot from 2021-09-12 18-59-20](https://user-images.githubusercontent.com/489601/132996215-71aad37b-df11-485a-92f8-c114752d152b.png)
![Screenshot from 2021-09-12 18-59-42](https://user-images.githubusercontent.com/489601/132996226-d4236298-0361-4bf6-9f18-5f74db96d3d2.png)

a1exsh commented:
Maybe we want to do something about that bottom right area on the "View World" dialog (in a separate PR):
![Screenshot from 2021-09-12 19-47-46](https://user-images.githubusercontent.com/489601/132997575-caf4a6bf-c5f5-4dce-8817-f1305127d8cd.png)

ihhub commented on src/fheroes2/gui/interface_border.cpp:
```diff
@@ -138,46 +210,95 @@ void Interface::GameBorderRedraw( const bool viewWorldMode )
-        srcrt.width = 3;
-        srcrt.height = isEvilInterface ? 15 : 20;
-        dstpt.y += 18;
+        fheroes2::Rect fixrt;
+        fheroes2::Point fixpt;
+        fixrt.x = 624;
+        fixrt.y = isEvilInterface ? 328 : 345;
+        fixrt.width = 3;
+        fixrt.height = isEvilInterface ? 15 : 20;
+        fixpt.x = dstpt.x;
+        fixpt.y = dstpt.y + 18;
```
Please use construtor calls as it'll be faster.

a1exsh replied:
> my main question here is do we render images on certain parts of monitor twice?

Only potentially really small parts maybe rendered twice, like up to `BORDERWIDTH × 6-8` pixels at a time.  We can further restrict the code to avoid that, but I don't think this makes a big difference performance-wise.

There are other things we could do to optimize rendering of repeated pattern, but I'd refrain from this, unless someone can demonstrate that this is critical.  Are we talking about specific platform, by the way, where this is known to slow things down?  I mean, on any modern hardware it should be really fast, as we don't do anything fancy here.

> Please use constructor calls as it'll be faster.

With optimizing compilers, there's probably no difference (especially since this is not even in a loop).  Yet the constructor is probably more compact/readable — so I'll give it a try.

a1exsh replied:
@ihhub please see 6f08ff8

ihhub replied:
We have players who own PS Vita or Orange Pi where FPS is much lower than on PCs. Also we're targeting mobile platform very soon where performance as well as the amount of energy are important.

a1exsh replied:
Will have a look how to avoid double-rendering.

ihhub reviewed (commented):
Hi @a1exsh , my main question here is do we render images on certain parts of monitor twice? This function is a bit performance important as it might be called many times during hero's movement causing slowdown in rendering. We must be sure that rendering (calling Blit) is done once per pixel/area.

ihhub commented:
> 
> 
> Maybe we want to do something about that bottom right area on the "View World" dialog (in a separate PR):
> ![Screenshot from 2021-09-12 19-47-46](https://user-images.githubusercontent.com/489601/132997575-caf4a6bf-c5f5-4dce-8817-f1305127d8cd.png)

Yes, we have to. Could you please create a separate issue for this so we won't forget?

ihhub commented:
@a1exsh , hmm... actually it's not good. We have to fix it here. View World shouldn't have extra line of interface.

a1exsh commented:
> @a1exsh , hmm... actually it's not good. We have to fix it here. View World shouldn't have extra line of interface.

I think this is an improvement over the current release where we have an unfinished border. Let's fix it separately.

ihhub commented:
> 
> 
> > @a1exsh , hmm... actually it's not good. We have to fix it here. View World shouldn't have extra line of interface.
> 
> I think this is an improvement over the current release where we have an unfinished border. Let's fix it separately.

In this case I'd suggest to use the background for date / army / resources. We already have an algorithm to scale it.

a1exsh commented:
re: view world
> In this case I'd suggest to use the background for date / army / resources. We already have an algorithm to scale it.

That's not that easy, it seems, as the View World, Mines, etc. all have custom resources complete with the text and buttons, not only background.  See `ICN::VIEW_ALL`, `VIEWRTFX`...

a1exsh commented:
> We must be sure that rendering (calling Blit) is done once per pixel/area.

So my idea is to introduce an additional overload for `Blit` (and `Copy`) functions, with the following signature:
```.cpp
    // draw a repeated pattern from in image area onto out image area
    void Blit( const Image & in, int32_t inX, int32_t inY, int32_t inWidth, int32_t inHeight,
               Image & out, int32_t outX, int32_t outY, int32_t width, int32_t height, int32_t offX = 0, int32_t offY = 0, bool flip = false );
```

It should fill the area of `width × height` pixels in the `out` image using the pattern, of potentially different width and height, from the area in the `in` image, and all that in one go.  The pattern would "wrap" around the source image area and get repeated in the destination area.

Using this function we can replace all the loops in the interface border rendering code with single calls to this new `Blit` variant, which itself ensures that each pixel in the destination area only rendered once.

What do you think?

ihhub commented:
> 
> 
> > We must be sure that rendering (calling Blit) is done once per pixel/area.
> 
> So my idea is to introduce an additional overload for `Blit` (and `Copy`) functions, with the following signature:
> 
> ```c++
>     // draw a repeated pattern from in image area onto out image area
>     void Blit( const Image & in, int32_t inX, int32_t inY, int32_t inWidth, int32_t inHeight,
>                Image & out, int32_t outX, int32_t outY, int32_t width, int32_t height, int32_t offX = 0, int32_t offY = 0, bool flip = false );
> ```
> 
> It should fill the area of `width × height` pixels in the `out` image using the pattern, of potentially different width and height, from the area in the `in` image, and all that in one go. The pattern would "wrap" around the source image area and get repeated in the destination area.
> 
> Using this function we can replace all the loops in the interface border rendering code with single calls to this new `Blit` variant, which itself ensures that each pixel in the destination area only rendered once.
> 
> What do you think?

This is very specific logic to have. I don't think we should put it to image.h/cpp files. We can put such function inside anonymous namespace within `interface_border.cpp` file.

a1exsh commented:
> This is very specific logic to have. I don't think we should put it to image.h/cpp files. We can put such function inside anonymous namespace within `interface_border.cpp` file.

I think this is not the only place where it can be used, see e.g: https://github.com/ihhub/fheroes2/blob/master/src/fheroes2/gui/interface_status.cpp#L412-L418

At the same time, I agree that we should probably not try to make existing `Blit` more complex than it already is. :sweat_smile: 
I will implement it for the border rendering first and we can see if we want to extract it later.

a1exsh commented:
@ihhub please see: https://github.com/ihhub/fheroes2/pull/4203/files#diff-5623c3e2b6aa90a870a922ea0dc08214bff39383b93ad544f102f003a58485c5R34-R65
No overlaps! :)

a1exsh commented:
I've extended the "View world" sprite by using the same approach as here: https://github.com/ihhub/fheroes2/blob/master/src/fheroes2/gui/interface_status.cpp#L412-L418

![Screenshot from 2021-09-18 19-51-02](https://user-images.githubusercontent.com/489601/133897926-3b661778-ecce-46f4-b8d6-2491d5262488.png)

a1exsh commented:
By the way, there are overlaps in the original code, e.g. the top-left corner is drawn both by top border and by left border rendering code.  See the following debug screen shot with `Blit` replaced with `DrawRect`:
![Screenshot from 2021-09-19 10-25-08](https://user-images.githubusercontent.com/489601/133920760-f33760f3-df92-4f35-b020-b56c49aca959.png)

I'd fix these as well then.

a1exsh commented:
Now the same, but without overlaps:
![image](https://user-images.githubusercontent.com/489601/133921723-c3ae43cd-7aed-4270-8349-b01f9125a897.png)

a1exsh commented:
Actually, while working on this I've realized that we wouldn't need to hide those embranchments (see #1286) if we painted them at the correct position in the first place.  This gives back the native look to the "icon border" on any resolution:

![image](https://user-images.githubusercontent.com/489601/133927255-d00f591c-2c54-464b-8fa2-6a1277703429.png)

What follows though, is even more differences between EVIL and GOOD border rendering code.  But now it looks solid, IMO.

ihhub commented on src/engine/image.cpp:
```diff
@@ -2094,4 +2094,37 @@ namespace fheroes2
         }
     }
+
+    void repeatPattern( const Image & in, int32_t inX, int32_t inY, int32_t inWidth, int32_t inHeight, Image & out, int32_t outX, int32_t outY, int32_t width,
```
In this case a double loop will be good enough. Yes, we'll have a bit slower code which is okay as it'll be more readable. Therefore, do we need this function in first place?

a1exsh replied:
I think such a function will be really useful, as I see many places in the code where the same is done ad-hoc with the manually-rolled loops.

I can move it back to where it's used here, for the border rendering, but I'd find a generic function handy in solving #1371 and #1423, for example.

a1exsh replied:
When it comes to handling the transformation layer: repeating it is easy, though really applying the transform on the output image is something I'd like avoid doing here, at least at first, as it is going increase complexity dramatically.  If needed this can be achieved by first repeating the pattern onto a temp. image, and then `Blit`-ing on the destination.

ihhub replied:
We can't guarantee that the destination image is a single layer image so calling Blit is safer rather than relying on an assumption of a single-layer image.

ihhub reviewed (changes requested):
Hi @a1exsh , I left multiple comments here. Would you mind please to take a look at them?

a1exsh commented on src/fheroes2/gui/interface_border.cpp:
```diff
@@ -29,6 +29,22 @@
+    inline void BlitR( const fheroes2::Image & in, int32_t x, int32_t y, fheroes2::Image & out, int32_t outX, int32_t outY, int32_t width, int32_t height )
+    {
+        fheroes2::Blit( in, x, y, out, outX, outY, width, height );
+        // fheroes2::DrawRect( out, fheroes2::Rect( outX, outY, width - 1, height - 1 ), 214 ); // DEBUG
```
Yes, this was for debug only.  I've found myself adding these debug rectangles again and again as I was working on the code, so I thought it makes sense to keep them commented out for the future.  I'm fine to remove them, of course.

a1exsh commented on src/engine/image.cpp:
```diff
@@ -2094,4 +2094,37 @@ namespace fheroes2
+            for ( ; imageOut != imageOutRepeatEnd; imageOut += inWidth ) {
+                memcpy( imageOut, imageIn, inWidth );
+            }
+            memcpy( imageOut, imageIn, restWidth );
```
On which platform do you see warnings?  I think `size_t` is the expected parameter type here?

a1exsh commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -41,6 +41,29 @@
+    {
+        const int32_t dstx = display.width() - viewWorldSprite.width() - BORDERWIDTH;
+        int32_t dsty = 2 * BORDERWIDTH + RADARWIDTH;
+        const int32_t cuty = 275;
```
Updated the name: `cutHeight`.  It's the height in the original image where we cut it and start repeating the background, if needed.

ihhub commented on src/fheroes2/gui/interface_border.cpp:
```diff
@@ -29,6 +29,42 @@
+        uint8_t * imageOut = out.image() + offsetOut;
+        uint8_t * const imageOutEnd = imageOut + height * widthOut;
+
+        for ( ; imageOut != imageOutEnd; imageOut += widthOut - width ) {
```
I'd still insist on calling Blit function here. Even though Display image is a single-layer image we have to call a function which can handle both cases while right now we aren't doing it.

a1exsh replied:
OK, we can do the same with `Blit`s.  Probably I was trying too hard to optimize it, after hearing your performance concerns. :sweat_smile:

a1exsh replied:
@ihhub re-wrote using `Blit`.  I don't see a performance impact on FullHD, so it should be good. :)  Btw, is there a reason we have to redraw the border on every frame of the hero moving on adventure map?  I don't see why that should be the case in the first place.

ihhub replied:
We shouldn't redraw the whole frame at first place. It comes from a bad design decision before us...

a1exsh commented:
@ihhub so far your comments were only about the code, but what do you think about the "new" look of the border?  I hope you also see it as an improvement. :)

ihhub commented:
> @ihhub so far your comments were only about the code, but what do you think about the "new" look of the border? I hope you also see it as an improvement. :)

I'll check the look today. Give me please some time🙂

a1exsh commented:
I will provide a recap of the changes shortly, as some differences may be too subtle to just see with a naked eye. ;)

a1exsh commented:
The following screenshots are taken at game resolution 700x525, because both width and height are not evenly divisible by `TILEWIDTH` (32 pixels).

1. GOOD interface.

   - Original:
     ![border_700x525_good_broken](https://user-images.githubusercontent.com/489601/135711226-f5346b84-5716-42a4-8cff-e882fab786f9.png)

   - Fixed:
     ![border_700x525_good_fixed](https://user-images.githubusercontent.com/489601/135711228-5575cb1c-8263-41b5-9699-c8701696879a.png)

   - Problems in the original rendering:
      a. Top/middle border junction:
          ![image](https://user-images.githubusercontent.com/489601/135711491-479a0dda-d03b-4b66-abaa-b3e556254f5c.png)
      b. Left/bottom border junction:
          ![image](https://user-images.githubusercontent.com/489601/135711520-ed7d6df9-f178-45ad-98c8-55986d21edcc.png)
      c. Middle and right/bottom border junction:
          ![image](https://user-images.githubusercontent.com/489601/135711542-ad0dcd04-15c6-44b9-bdd3-5500a8e9bfe1.png)

   - How the problems are fixed:

      - Only use as many central tiles in the top and bottom border as fits in the extra screen width:
         - ![image](https://user-images.githubusercontent.com/489601/135711727-341c968c-b394-4ed4-b351-0b2f5f90a551.png)
         - ![image](https://user-images.githubusercontent.com/489601/135711894-95065657-d761-49af-84f6-a4b70c55448c.png)

      - Fill the rest of the space by repeating a narrow slice of the border.
        In the GOOD top and bottom border pattern, the only piece that seems to fit nicely is the straight part of the left/right wing (marked with red in the above image).

      - Also, in the left border, use as many repeating tiles as fits in the extra height and fill the rest with a repeating slice:
        Original: ![image](https://user-images.githubusercontent.com/489601/135711993-ebcd8a58-b697-4e94-a646-e3810fb6694c.png) Extended: ![image](https://user-images.githubusercontent.com/489601/135711981-0f291ed5-0e97-4c29-9a44-b9dac23b7d9b.png)

      - The same is applied for the right border.

      - The middle border is particularly tricky on the GOOD interface due to the leaves which are not that easy to repeat, but the result looks decent, IMO:
         - ![image](https://user-images.githubusercontent.com/489601/135712226-2a367814-c21b-41d0-a73a-2e654d7388bf.png)

      - As a bonus, we do not have to hide the branching when rendering normal border (we only do that for the "View World" mode):
        - Original: ![image](https://user-images.githubusercontent.com/489601/135712293-24e4fdc1-483f-4644-b0c1-e94c36c4eeda.png)
        - Fixed: ![image](https://user-images.githubusercontent.com/489601/135712299-62ca5d62-421c-4c8d-a675-a02edf3e8c4f.png)

2. EVIL interface.
   - Original:
     ![border_700x525_evil_broken](https://user-images.githubusercontent.com/489601/135711239-7df6357e-7955-447b-bfa9-ee2c36efd430.png)
   - Fixed:
     ![border_700x525_evil_fixed](https://user-images.githubusercontent.com/489601/135711240-482347d1-c600-4e4d-bf03-d32119f829e3.png)

   - Mostly the same problems are present here, and we fix most of them the same way:

      - The top boder:
         - ![image](https://user-images.githubusercontent.com/489601/135712374-46d50d95-a5da-4665-b1c0-c188204eefa8.png)
         - ![image](https://user-images.githubusercontent.com/489601/135712389-5f93369f-5c3c-493c-aab4-f20e44bd1caf.png)

      - Left/middle/right border:
        Original: ![image](https://user-images.githubusercontent.com/489601/135712467-315f63cf-ca4e-44f0-a677-956eb4f3bf63.png) Extended: ![image](https://user-images.githubusercontent.com/489601/135712489-9ad7a959-b481-4ae7-b41a-bb23bb479b64.png)

      - "Icon" border:
        - Original: ![image](https://user-images.githubusercontent.com/489601/135712530-7c2db2cb-5e8a-4740-96d3-dd0bf814e414.png)
        - Fixed: ![image](https://user-images.githubusercontent.com/489601/135712562-3e1a7cb4-1a3c-4296-b694-cc8c993c619e.png)

      - While in the GOOD interface the pattern in the middle and right border is the same before and after the branching (that wooden column), in the EVIL interface the pattern is different.  I had to find another piece of 32px to repeat and this one was fitting nicely:
         - ![image](https://user-images.githubusercontent.com/489601/135712702-ae894c94-ebd1-4b62-825e-3d118d28c0b9.png)

      - Finally, EVIL would not be evil, if not for all the small asymmetries that it comes with. ;)
        The bottom border doesn't have the 32px-wide central piece to repeat, which led to this problem in the original rendering:

         - ![image](https://user-images.githubusercontent.com/489601/135712819-107ce626-8791-4036-a698-65348ee1ddf3.png)
         - I've fixed this by using a different size of the repeating pattern for the bottom, in case of the EVIL interface.  Here we only repeat the 7px-wide bone: 
![image](https://user-images.githubusercontent.com/489601/135712850-cdea47ad-6541-4eae-b90f-47087c1434ac.png)

ihhub commented:
Hi @a1exsh , these are excellent findings. The original code was written just to make the game work without paying attention to the correctness of the implementation.

ihhub commented:
Hi @a1exsh , overall looks better than before.

@Branikolog and @LeHerosInconnu, could you please express your opinion about the look?

Branikolog commented:
Hello everyone.
I didn't notice any flaws in current PR implementation on any resolution my PC supports.
Anyway, @a1exsh , you've done a tremendous work exploring and fixing this frames. I really appreciate your contribution and respect your skills. :)

LeHerosInconnu commented:
Hello everyone,

> Hi @a1exsh , overall looks better than before.
> 
> @Branikolog and @LeHerosInconnu, could you please express your opinion about the look?

So I tested different resolutions and did not detect any problems so far.
Except maybe... no just kidding. :D
Very good work! Thanks. :)

a1exsh commented on src/fheroes2/gui/interface_border.cpp:
```diff
@@ -29,6 +29,40 @@
+        const int32_t countY = height / inHeight;
+        const int32_t restWidth = width % inWidth;
+        const int32_t restHeight = height % inHeight;
+        int32_t x;
+        int32_t y;
```
These are intentionally declared outside of the loop, because we use the values after the loop is over to position the last piece of the row/column.  I can try to put them inside the loop, but I fear the code will be lot less cleaner with that.

ihhub replied:
At least we can put `x` within the loop :)

ihhub replied:
Ah, yes. Now I got it. Let's leave it as it is.

a1exsh replied:
Ah, actually, it's not as bad as I thought: https://github.com/a1exsh/fheroes2/commit/7de857706844939ff8bab8af00b2bf0c31cc458d

I see that you've merged already.  I can make a quick follow-up PR, if you like. ;)

a1exsh replied:
@ihhub Feel free to merge or close: https://github.com/ihhub/fheroes2/pull/4312

ihhub commented:
@a1exsh , a huge thanks for this great work!

a1exsh commented:
@ihhub you might now want to update one of the images from the project page, as it has exactly the same border problem at 800x600 resolution:
![image](https://user-images.githubusercontent.com/489601/135762874-8cc53f09-e67c-430e-8ae3-5c340c035106.png)
